src/shogun/mathematics/linalg/linalg.h

```diff
@@ -35,40 +35,40 @@
 #include <shogun/lib/config.h>
 
 /**
- * Just include this file to use in your applications.
+ * Just include this file to use in your applications (in the cpp)
  */
 
-/**
- * This namespace contains all linear algebra specific modules and operations.
- * The rest of the library is available only when the one of the supported
- * backend exists in the local system
- */
-
-#ifdef HAVE_LINALG_LIB
 namespace shogun
 {
 
+/**
+ * This namespace contains all linear algebra specific modules and operations
+ * for which we (may) rely on multiple implementation, either native or
+ * making use of some supported third party external linear algebra libraries.
+ */
 namespace linalg
 {
 
 /**
  * Developer's Note :
- * - Changing the default backend would just require to change it in the following
- *   enum
+ * - Changing the default backend would just require to change it in the
+ *   Backend enum
  * - Please use the same names as HAVE_<BACKEND> macros (e.g. HAVE_EIGEN3)
+ * - Please refer to the developer's wiki (link) for the design and structure
+ *   of internal linalg module
  */
 
 /**
  * @brief
  * All currently supported linear algebra backend libraries, with a default
  * backend, which will be used for all the tasks if any particular backend is
- * not set explicitly via cmake options.
+ * not set explicitly via cmake options. A Native backend is also defined to
+ * accommodate Shogun's native implementation of some operations.
  *
  * The enum defines these backends in order of priority as default backend, as
- * in, first defined one will be used as default
- *
- * Note - Currently EIGEN3 is the default (if its available)
+ * in, first defined one will be used as default.
  *
+ * Note - Currently EIGEN3 is the default (if it is available).
```

Comment:
@iglesias actually we don't want to implement _everything_ natively, e.g. we most certainly don't want to write our own self adjoint eigensolver or SVD/QR/Cholesky. `HAVE_LINALG_LIB` guard is useful for those methods which are available _only_ when we _do_ have an external linalg lib. In linalg modules we're guarding those operations with `HAVE_LINALG_LIB` which do not have their native implementations (this modification is done in this patch only). I think it is still useful. A use-case is when we just want the result for some of these operations and don't wanna bother about which backend was used (i.e. relying on global backend setup).

In CustomKernel.cpp I'm explicitly using eigen3 backend because when I added this, ViennaCL specializations were not added till then for those methods. We can most certainly do native implementation for `sum` methods, and, in that case, then we can get rid of the guard in CustomKernel. Otherwise, as of now, we can get rid of `HAVE_EIGEN3` guard and remove the specific backend Eigen3 in those `sum` methods - it will then use Eigen3/ViennaCL - whichever is available.